### PR TITLE
Fix import error messages

### PR DESCRIPTION
**Pull request by Frassle**
# Description

Fixes https://github.com/pulumi/pulumi/issues/13062

## Checklist

- [ ] I have run `make tidy` to update any new dependencies
- [ ] I have run `make lint` to verify my code passes the lint check
  - [ ] I have formatted my code using `gofumpt`

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

File: pkg/cmd/pulumi/import.go [hunk 1]
```diff
@@ -540,8 +543,17 @@ func newImportCmd() *cobra.Command {
...
+					return result.Errorf("an inline resource must be specified if no converter or import file is used, missing type, name, and id")
+				}
+				if len(args) == 1 {
+					cmd.Help()
+					return result.Errorf("an inline resource must be specified if no converter or import file is used, missing name and id")
+				}
+				if len(args) == 2 {
+					cmd.Help()
+					return result.Errorf("an inline resource must be specified if no converter or import file is used, missing id")
 				}
```
Comment (abhinav) [inline pkg/cmd/pulumi/import.go:557]: This can be simplified:

```go
fields := []string{"type", "name", "id"}
if len(args) > len(fields) {
  return result.Errorf("unexpected arguments: %q", args[len(fields):])
}
missing := fields[:len(args)]
if len(missing) > 0 {
  return result.Errorf("missing fields: %q", missing)
}

// ...
```

(with appropriate tweaks for help messages, etc.)

**Review by abhinav — COMMENTED** [review]
I think an accompanying test for the error messages returned when the command is invoked with bad arguments wouldn't be too difficult here.

File: pkg/cmd/pulumi/import.go (hunk 1 above)
Comment (Frassle) [reply pkg/cmd/pulumi/import.go:557]: I think I actually prefer the long form. It's a little repetitive but it's simple and it's a clear user facing error. I wouldn't want to shorten this to just "missing fields: id". 

We could have something like "an inline resource must be specified if no converter or import file is used, missing: type, name, id" but then we lose the nice grammar at the end and doing that bit correctly with code seems more complex than helpful.

I will pull the first part of the message out to a common var though, no need to write that out three times.

File: pkg/cmd/pulumi/import.go [hunk 2]
```diff
@@ -540,8 +543,18 @@ func newImportCmd() *cobra.Command {
...
+					contract.IgnoreError(cmd.Help())
+					return result.Errorf(msg + "type, name, and id")
+				}
+				if len(args) == 1 {
+					contract.IgnoreError(cmd.Help())
+					return result.Errorf(msg + "name and id")
+				}
+				if len(args) == 2 {
+					contract.IgnoreError(cmd.Help())
+					return result.Errorf(msg + "id")
```
Comment (abhinav) [inline pkg/cmd/pulumi/import.go:557]: Ack, although for the manual variant, consider a switch statement?
```suggestion
				switch len(args) {
				case 0:
					contract.IgnoreError(cmd.Help())
					return result.Errorf(msg + "type, name, and id")
				case 1:
					contract.IgnoreError(cmd.Help())
					return result.Errorf(msg + "name and id")
				case 2:
					contract.IgnoreError(cmd.Help())
					return result.Errorf(msg + "id")
```

**Review by abhinav — APPROVED** [review]

File: pkg/cmd/pulumi/import.go (hunk 2 above)
Comment (Frassle) [reply pkg/cmd/pulumi/import.go:557]: Sure

**Comment by Frassle** [comment]
> I think an accompanying test for the error messages returned when the command is invoked with bad arguments wouldn't be too difficult here.

It's actually a surprising amount of work. We can't easily test against the `cobra.Cmd` directly because of the various immediate exit hooks we apply for correct CLI functioning. So I'm going to have to do the standard trick of splitting this into it's own testable struct as we've done for other commands. 
That's a large diff though, so I'll open a separate PR for that then come back to this one.

**Comment by Frassle** [comment]
> That's a large diff though, so I'll open a separate PR for that then come back to this one.

Urgh it's a horrible change, we need the cobra.Cmd to call Help() on. But can't execute the Cmd directly because it quick exits.
I'm inclined to say we merge this as is, I have manually checked these errors look ok. 

I'll give some thinking to how we can test commands of this shape.

**Comment by Frassle** [comment]
bors merge
